File: 프로그래머스/Level 2/프렌즈4블록-ex.py
def find_blocks(i, j, board):
    m, n = len(board), len(board[0])
    same_blocks = [(i, j)]
    
    for dx, dy in [(0, 1), (1, 0), (1, 1)]:
        nx, ny = i + dx, j + dy
        if nx < 0 or nx >= m or ny < 0 or ny >= n or board[i][j] != board[nx][ny]:
            return []
        same_blocks.append((nx, ny))
    
    return same_blocks

def remove_blocks(removed_blocks, board):
    removed_block_num = 0 
    for x, y in removed_blocks:
        board[x][y] = 0 
        removed_block_num += 1
    return removed_block_num 

def fall_blocks(m, n, board): 
    for j in range(n): 
        for i in range(m - 1, 0, -1): 
            prev = i - 1
            while prev >= 0 and board[i][j] == 0: 
                board[i][j], board[prev][j] = board[prev][j], board[i][j]
                prev -= 1

def solution(m, n, board):
    board = [list(row) for row in board]  # 문자열 리스트를 2D 문자열 배열로 변환
    removed_block_num = 0 
    
    while True: 
        is_found = False 
        removed_blocks = set() 
        
        for i in range(m): 
            for j in range(n): 
                if board[i][j] == 0:
                    continue 
                found_blocks = find_blocks(i, j, board)
                if found_blocks:
                    is_found = True 
                    removed_blocks.update(found_blocks)
        
        if not is_found: 
            break 
            
        removed_block_num += remove_blocks(removed_blocks, board)
        fall_blocks(m, n, board)
    
    return removed_block_num

File: 프로그래머스/Level 2/프렌즈4블록-ex.java
import java.util.List;
import java.util.ArrayList;

class Solution {
    private static final int[][] nears = new int[][]{{0, 1}, {1, 0}, {1, 1}};
    public int solution(int m, int n, String[] board) {
        String[][] newBoard = new String[m][n];
        int removedBlockNum = 0;
        boolean isFound = true;
        
        copyBoard(board, newBoard);
        
        while (isFound) {
            isFound = false;
            List<int[]> removedBlocks = new ArrayList<>();
            List<int[]> foundBlocks;
            
            for (int i = 0; i < m; i++) {
                for (int j = 0; j < n; j++) {
                    if (newBoard[i][j].equals("0")) continue;
                    
                    foundBlocks = findBlocks(i, j, newBoard);
                    if (!foundBlocks.isEmpty()) {
                        isFound = true;
                        removedBlocks.addAll(foundBlocks);
                    }
                }
            }
            
            if (!isFound) break;
            
            removedBlockNum += removeBlocks(removedBlocks, newBoard);
            
            fallBlocks(m, n, newBoard);
        }
        
        return removedBlockNum;
    }
    
    private void copyBoard(String[] source, String[][] dest) {
        for (int i = 0; i < source.length; i++) {
            destination[i] = source[i].split("");
        }
    }
    
    private List<int[]> findBlocks(int x, int y, String[][] board) {
        List<int[]> sameBlocks = new ArrayList<>();
        int m = board.length;
        int n = board[0].length;
        
        sameBlocks.add(new int[]{x, y});
        
        for (int i = 0; i < nears.length; i++) {
            int nx = x + nears[i][0];
            int ny = y + nears[i][1];
            
            if (nx < 0 || nx >= m || ny < 0 || ny >= n || 
               !board[x][y].equals(board[nx][ny]))
                return new ArrayList<int[]>();
            else if (board[x][y].equals(board[nx][ny])) 
                sameBlocks.add(new int[]{nx, ny});
        }
        
        return sameBlocks;
    }
    
    private int removeBlocks(List<int[]> removedBlocks, String[][] board) {
        int m = board.length;
        int n = board[0].length; 
        int removedBlockNum = 0;
        int emptyNum = 0;
        
        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                if (board[i][j].equals("0")) emptyNum++;
            }
        }
        
        
        for (int[] block : removedBlocks) {
            board[block[0]][block[1]] = "0";
        }
        
        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                if (board[i][j].equals("0")) removedBlockNum++;
            }
        }
        
        return removedBlockNum - emptyNum;
    }
    
    private void fallBlocks(int m, int n, String[][] board) {
        for (int j = 0; j < n; j++) {
            for (int i = m - 1; i > 0; i--) {
                int prev = i - 1;
                while (prev >= 0 && board[i][j].equals("0")) {
                    board[i][j] = board[prev][j];
                    board[prev][j] = "0";
                    prev--;
                }
            }
        }
    }
}